// mq-cli/src/commands/login.js
const {Command, flags} = require('@oclif/command');
const os = require('os');
var fs = require('fs');
var cli = require('cli-ux');
var firebase = require('firebase');
var client = require('../lib/client');

var config_dir = os.homedir() + "/.mq";
var firebase_app = null;

var firebase_config = {
  apiKey: "AIzaSyDPMhe2jLsYpd_kBHCSwVM9vxLxGIp19Us",
  authDomain: "moquality-inc.firebaseapp.com",
  databaseURL: "https://moquality-inc.firebaseio.com",
  storageBucket: "moquality-inc.appspot.com",
};

function firebase_login_with_email(email, password) {
  firebase_app = firebase.initializeApp(firebase_config);
  return firebase.auth().signInWithEmailAndPassword(email, password);
}

function firebase_login_with_token(token) {
  firebase_app = firebase.initializeApp(firebase_config);
  return firebase.auth().signInWithCustomToken(token);
}

function log_in(email, password, api_key) {
  return new Promise(function (resolve, reject) {
    if (api_key) {
      client.get('user/token', {
          headers: {'X-API-KEY': api_key}
      }).then(function (response) {
        return resolve(firebase_login_with_token(response.data.token));
      }).catch(function (err) {
        reject(new Error(err.message));
      });
    } // login with api_key
    else if (email && password) {
      return resolve(firebase_login_with_email(email, password));
    } // login with email
    else {
      reject(new Error('Give me your api_key!'));
    }
  });
}

class LoginCommand extends Command {
  async run() {
    const {flags} = this.parse(LoginCommand)
    var email = flags.email;
    var password = flags.password;
    var api_key = flags.api_key;

    if (api_key == null) {
      if (email == null) {
        email = await cli.cli.prompt('What is MoQuality email?')
      }
      if (password == null) {
        password = await cli.cli.prompt('What is your password?', {type: 'hide'})
      }
    }

    log_in(email, password, api_key).then(function() {
      var data = {
          'api_key': firebase_app.auth().currentUser.uid,
          'email': firebase_app.auth().currentUser.email,
      };
      if (!fs.existsSync(config_dir)){
        fs.mkdirSync(config_dir);
      }
      fs.writeFile(config_dir + "/config.json", JSON.stringify(data), function(err) {
          if (err) reject(err);
          else console.log(`Welcome ${firebase_app.auth().currentUser.displayName}`);
      });
    }).catch(function(err) {
        console.log(err.message);
    });

  }
}

LoginCommand.description = `Login to MoQuality`

LoginCommand.flags = {
  email: flags.string({char: 'e', description: 'email of user'}),
  password: flags.string({char: 'p', description: 'password of user'}),
  api_key: flags.string({char: 'a', description: 'api_key of user'}),
}

module.exports = LoginCommand
