// mq-cli/src/lib/config.js

var fs = require('fs');
const os = require('os');
var firebase = require("firebase");
var client = require('./client');

var firebase_config = {
    apiKey: "AIzaSyDPMhe2jLsYpd_kBHCSwVM9vxLxGIp19Us",
    authDomain: "moquality-inc.firebaseapp.com",
    databaseURL: "https://moquality-inc.firebaseio.com",
    storageBucket: "moquality-inc.appspot.com",
  };

var firebase_app = null;

function firebase_login(token) {
    firebase_app = firebase.initializeApp(firebase_config);
    return firebase.auth().signInWithCustomToken(token);
}

module.exports = function () {
    return new Promise(function(resolve, reject) {
        var config_path = os.homedir() + '/.mq/config.json';
        if (fs.existsSync(config_path)) {
            config = JSON.parse(fs.readFileSync(config_path, 'utf8'));
            client.get('user/token', {
                headers: {'X-API-KEY': config['api_key']}
            }).then(function (response) {
                config['token'] = response.data.token;
                firebase_login(config['token']).then(function() {
                    config['firebase_app'] = firebase_app;
                    resolve(config);
                });
            }).catch(function (err) {
                reject(err);
            });
        } else {
            console.log('Login to continue');
            process.exit();
        }
    });
}
